rust/rsmgp-sys/src/value/mod.rs

```diff
@@ -0,0 +1,326 @@
+use std::ffi::{CStr, CString};
+
+use crate::context::*;
+use crate::edge::*;
+use crate::mgp::*;
+use crate::path::*;
+use crate::result::*;
+use crate::vertex::*;
+// Required here, if not present tests linking fails.
+#[double]
+use crate::mgp::ffi;
+use mockall_double::double;
+
+// mgp_value used to return data is non-const, owned by the module code. Function to delete
+// mgp_value is non-const.
+// mgp_value containing data from Memgraph is const.
+//
+// `make` functions return non-const value that has to be deleted.
+// `get_property` functions return non-const copied value that has to be deleted.
+// `mgp_property` holds *const mgp_value.
+//
+// Possible solutions:
+//   * An enum containing *mut and *const can work but the implementation would also contain
+//   duplicated code.
+//   * A generic data type seems complex to implement https://stackoverflow.com/questions/40317860.
+//   * Holding a *const mgp_value + an ownership flag + convert to *mut when delete function has to
+//   be called.
+//   * Hold only *mut mgp_value... as soon as there is *const mgp_value, make a copy (own *mut
+//   mgp_value). The same applies for all other data types, e.g. mgp_edge, mgp_vertex.
+//   mgp_value_make_vertex accepts *mut mgp_vartex (required to return user data), but data from
+//   the graph is all *const T.
+
+/// Useful to hold user created `mgp_value` but also to store values coming from Memgraph, e.g.,
+/// property `mgp_value`s.
+#[derive(Debug)]
+pub struct MgpValue {
```

Comment:
As commented before, this might be renamed to Value, and that would be the only object named Value in the whole module. Example of C++ wrapper: https://github.com/memgraph/mage/blob/T027-MAGE_cpp-wrapper/cpp/mg_procedure/mgp.hpp#L664

Reply:
Yep, good point, it's 100% possible to have one `Value` object, but having 2 allows a couple of nice things:
1. Pure `MgpValue` containing just a pointer can be used internally to destroy `mgp_value` when needed.
2. Pure Rust `Value` could be used in [pattern matching](https://doc.rust-lang.org/book/ch06-00-enums.html).
```
if let Value::Int(value) = value { // do something with int */ }
else if let Value::String(value) = value { /* do something with string */ }
```
I think having both, will simplify the Rust code, but yea, users might be confused with two similar types.
I'll definitely think about what to do with the "value" type.

Reply:
I think I'll keep both of these objects for now. @jmatak let me know if that still doesn't make sense :smile: